### PR TITLE
Plotting normalization: ignore nodata pixels

### PR DESCRIPTION
**Pull request by adamjstewart**
For images containing nodata pixels, these zeros completely throw off our math during normalization. For reproducibility, the below figures are from our I/O Bench dataset.

### Before

before

### After

after

## AI Usage Disclosure

- [x] 🟢 **No AI usage**: written by humans, for humans
- [ ] 🟡 **AI-assisted**: AI helped with the coding, but I understand every line
- [ ] 🔴 **AI-generated**: AI did everything, review with caution

### CODE REVIEW

**Comment by adamjstewart** [comment]
Question: should we error on all zero images (like the one failing test) or gracefully return all zeros?

**Comment by robmarkcole** [comment]
@adamjstewart which failing test? An all zero image could be confusing for users, an error feels appropriate

**Comment by adamjstewart** [comment]
It was the Copernicus-Pretrain test that was failing, I guess our fake image for that one is all zeros.

All zero image isn't uncommon (especially because we currently sample from nodata regions). I didn't want an error in the middle of an epoch (we do plot during validation in our trainers).

**Comment by robmarkcole** [comment]
Maybe just a warning/info on plotting then, dont want to terminate training over that

**Comment by calebrob6** [comment]
You can't assume that nodata is 0 and doing a mask like this can take a relatively long time at S2 scene level. Would suggest allowing a `nodata_mask` or something if you want to subset which pixels the stats are computed over.

Also, this is why fixed RGB normalization for different platforms is useful. E.g. if I have the S2 RGB bands I can just divide by 3000 or 4000 and clip to get something that looks good regardless of nodata, scene composition, etc.

**Comment by adamjstewart** [comment]
> You can't assume that nodata is 0

This is a valid point. We've talked about adding a `nodata` parameter to `RasterDataset` (default to 0) that users could override. For XarrayDataset, we already hardcode `nodata=0`, but I don't think we yet do this in `RasterDataset`.

> doing a mask like this can take a relatively long time at S2 scene level.

This is just a random example, I don't expect users to do this very often. This method is only used during plotting, never during data augmentation.

> Would suggest allowing a nodata_mask or something if you want to subset which pixels the stats are computed over.

I think this would just move the complexity elsewhere, right?

**Comment by calebrob6** [comment]
> I think this would just move the complexity elsewhere, right?

Yeah I guess this doesn't help as its used in all the plot methods -- each would just need to compute masks there. My thought was that making it explicit forces users to realize that there is a mask happening. If this is just on patches it won't be an issue.


We can propagate nodatas later. Fine to merge

**Review by calebrob6 — DISMISSED** [review]

File: torchgeo/datasets/utils.py [hunk 1]
```diff
@@ -818,12 +819,18 @@ def percentile_normalization(
...
     .. versionadded:: 0.2
     .. versiondeprecated:: 0.10
     """
+    if not np.any(img):
+        return img
+
     assert lower < upper
-    lower_percentile = np.percentile(img, lower, axis=axis)
-    upper_percentile = np.percentile(img, upper, axis=axis)
+    lower_percentile = np.percentile(img[img != 0], lower, axis=axis)
```
Comment (robmarkcole) [inline torchgeo/datasets/utils.py]: Could we not pass in an arg for the nodata value, default to 0?

**Review by robmarkcole — DISMISSED** [review]

File: torchgeo/datasets/utils.py (hunk 1 above)
Comment (adamjstewart) [reply torchgeo/datasets/utils.py]: Done, hopefully this also addresses one of @calebrob6's concerns as well.

**Review by robmarkcole — APPROVED** [review]
